Core/src/Propagator/detail/CovarianceEngine.cpp

```diff
@@ -44,41 +47,36 @@ FreeToBoundMatrix freeToCurvilinearJacobian(const Vector3& direction) {
   const double x = direction(0);  // == cos(phi) * sin(theta)
   const double y = direction(1);  // == sin(phi) * sin(theta)
   const double z = direction(2);  // == cos(theta)
-  // can be turned into cosine/sine
-  const double cosTheta = z;
-  const double sinTheta = std::hypot(x, y);
-  const double invSinTheta = 1. / sinTheta;
-  const double cosPhi = x * invSinTheta;
-  const double sinPhi = y * invSinTheta;
   // prepare the jacobian to curvilinear
   FreeToBoundMatrix jacToCurv = FreeToBoundMatrix::Zero();
-  if (std::abs(cosTheta) < s_curvilinearProjTolerance) {
+  if (std::abs(z) < s_curvilinearProjTolerance) {
+    auto [cosPhi, sinPhi, cosTheta, sinTheta] =
+        VectorHelpers::evaluateTrigonomics(direction);
     // We normally operate in curvilinear coordinates defined as follows
-    jacToCurv(0, 0) = -sinPhi;
-    jacToCurv(0, 1) = cosPhi;
-    jacToCurv(1, 0) = -cosPhi * cosTheta;
-    jacToCurv(1, 1) = -sinPhi * cosTheta;
-    jacToCurv(1, 2) = sinTheta;
+    jacToCurv(eBoundLoc0, eFreePos0) = -sinPhi;
+    jacToCurv(eBoundLoc0, eFreePos1) = cosPhi;
+    // jacToCurv(eBoundLoc0, eFreePos2) = 0;
+    jacToCurv(eBoundLoc1, eFreePos0) = -cosPhi * cosTheta;
+    jacToCurv(eBoundLoc1, eFreePos1) = -sinPhi * cosTheta;
+    jacToCurv(eBoundLoc1, eFreePos2) = sinTheta;
   } else {
     // Under grazing incidence to z, the above coordinate system definition
     // becomes numerically unstable, and we need to switch to another one
     const double c = std::hypot(y, z);
     const double invC = 1. / c;
-    jacToCurv(0, 1) = -z * invC;
-    jacToCurv(0, 2) = y * invC;
-    jacToCurv(1, 0) = c;
-    jacToCurv(1, 1) = -x * y * invC;
-    jacToCurv(1, 2) = -x * z * invC;
+    // jacToCurv(eBoundLoc0, eFreePos0) = 0;
+    jacToCurv(eBoundLoc0, eFreePos1) = -z * invC;
+    jacToCurv(eBoundLoc0, eFreePos2) = y * invC;
+    jacToCurv(eBoundLoc1, eFreePos0) = c;
+    jacToCurv(eBoundLoc1, eFreePos1) = -x * y * invC;
+    jacToCurv(eBoundLoc1, eFreePos2) = -x * z * invC;
   }
   // Time parameter
-  jacToCurv(5, 3) = 1.;
+  jacToCurv(eBoundTime, eFreeTime) = 1.;
   // Directional and momentum parameters for curvilinear
-  jacToCurv(2, 4) = -sinPhi * invSinTheta;
-  jacToCurv(2, 5) = cosPhi * invSinTheta;
-  jacToCurv(3, 4) = cosPhi * cosTheta;
-  jacToCurv(3, 5) = sinPhi * cosTheta;
-  jacToCurv(3, 6) = -sinTheta;
-  jacToCurv(4, 7) = 1.;
+  jacToCurv.block<2, 3>(eBoundPhi, eFreeDir0) =
+      freeToSphericalDirectionJacobian(direction);
```

Comment:
This reruns the trigonomics internally again doesn't it? Maybe that was the reason why this was previously manually inlined?

Reply:
I hoped with the inlining of `sphericalToFreeDirectionJacobian` the second call would be removed by the compiler but I don't know